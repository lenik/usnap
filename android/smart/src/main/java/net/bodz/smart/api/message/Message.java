package net.bodz.smart.api.message;

/**
 * A Comment message.
 */
public class Message {

}
