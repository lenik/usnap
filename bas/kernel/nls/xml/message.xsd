<?xml version="1.0" encoding="UTF-8"?>
<!-- edited with XMLSpy v2005 U (http://www.xmlspy.com) by Xmlspy (Ru-Board) -->
<xs:schema xmlns:msg="http://www.bodz.net/xml/current/bas.k.nls.x.message" xmlns:xs="http://www.w3.org/2001/XMLSchema" targetNamespace="http://www.bodz.net/xml/current/bas.k.nls.x.message" elementFormDefault="qualified" attributeFormDefault="unqualified">
	<xs:element name="message-set">
		<xs:annotation>
			<xs:documentation>bas.k.nls.x.message</xs:documentation>
		</xs:annotation>
		<xs:complexType>
			<xs:sequence>
				<xs:element name="message">
					<xs:annotation>
						<xs:documentation>Use xml:lang to indicate what language is used. </xs:documentation>
					</xs:annotation>
					<xs:complexType mixed="true">
						<xs:choice minOccurs="0" maxOccurs="unbounded">
							<xs:element name="variable">
								<xs:annotation>
									<xs:documentation>Variable insert</xs:documentation>
								</xs:annotation>
								<xs:complexType>
									<xs:attribute name="name" type="xs:string" use="required"/>
								</xs:complexType>
							</xs:element>
						</xs:choice>
						<xs:attribute name="name" type="xs:string" use="required"/>
						<xs:attribute name="comment" type="xs:string"/>
					</xs:complexType>
				</xs:element>
			</xs:sequence>
			<xs:attribute name="name" type="xs:string"/>
		</xs:complexType>
	</xs:element>
</xs:schema>
