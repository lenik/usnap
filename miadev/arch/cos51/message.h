#ifndef COS51_MESSAGE_H
#define COS51_MESSAGE_H

#include <stdbool.h>

#include "types.h"

typedef unsigned char m_t;

#define _M_VOID(x)      (0x00 | (x))
#define _M_B(x)         (0x40 | (x))
#define _M_W(x)         (0x80 | (x))
#define _M_BB(x)        (0xA0 | (x))
#define _M_BW(x)        (0xC0 | (x))

#define M_DEV_ATTACH    _M_B(1)
#define M_DEV_DETACH    _M_B(2)
#define M_POWER _M_B(3)

#define M_KEYDOWN       _M_B(10)
#define M_KEYUP         _M_B(11)
#define M_KEY           _M_B(12)
#define M_MOUSEDOWN     _M_B(13)
#define M_MOUSEUP       _M_B(14)
#define M_CLICK         _M_B(15)
#define M_DBLCLICK      _M_B(16)

#define M_RECV          _M_BB(1)
#define M_RECVL         _M_BW(1)
#define M_SENT          _M_BB(1)

//SDCC doesn't expand the macro until it's used.
//#undef _M_VOID
//#undef _M_B
//#undef _M_W
//#undef _M_BB
//#undef _M_BW

typedef struct _PollEntry {

    /**
     * @return non-zero if there has event been enqueued.
     */
    byte (*poll)();

    int pollInterval;

} PollEntry;

extern int pollCount;
extern const PollEntry *pollTable;

/**
 * @return non-zero if the message is handled.
 */
typedef byte (*MessageHandler)(m_t message, byte param, word ndata) __reentrant;

// extern int handlerCount;
// SDCC-BUGFIX: declare/define type mismatches: (in large-mode only)
// extern const MessageHandler *handlerTable;

#define INITMQ(PTAB, HTAB, MS) \
    initMQ(sizeof(PTAB) / sizeof(__code void *), (PTAB), \
            sizeof(HTAB) / sizeof(__code void *), (HTAB), \
            (MS))

void initMQ(
        int pollCount, const PollEntry *pollTable,
        int handlerCount, const MessageHandler *handlerTable,
        int mbufSize
);

typedef __xdata byte *mbuf_t;

extern mbuf_t mbuf;
extern __xdata int statMTotal;
extern __xdata int statMDrops;

void enqueue_void(m_t message);
void enqueue_b(m_t message, byte param);
void enqueue_bb(m_t message, byte param, byte byteData);
void enqueue_w(m_t message, word data);
void enqueue_bw(m_t message, byte param, word data);

/**
 * @return false if the queue is empty.
 */
bool dequeue();

/**
 * The main function should as follow:
 *
 * void timer() interrupt 1 { pollEvents(); }
 * void main() { INITMQ(...); while (1) dequeue(); }
 */
void pollEvents();

#endif
