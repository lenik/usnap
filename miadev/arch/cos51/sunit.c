#include <stdio.h>
#include <8051.h>

#define SUNIT_IMPL
#include "sunit.h"

extern __xdata byte sunitBuffer[];
extern byte sunitBufferSize;

static int _start = 0;
static int _end = 0;

void putchar(char c) {
    while (!TI)
        ;
    TI = 0;
    SBUF = c;
}

char getchar() {
    byte c;
    while (_start == _end)
        ;
    c = sunitBuffer[_start++];
    _start %= sunitBufferSize;
    return (char) c;
}

void sunitSerialProc()
__interrupt 4 {
    if (RI) {
        if (_start <= _end) {
            sunitBuffer[_end++] = SBUF;
            _end %= sunitBufferSize;
        } else if (_end + 1 < _start) {
            sunitBuffer[_end++] = SBUF;
        } else {
            die("recvbuf full");
        }
        RI = 0;
    }
}

void vmstop() {
__asm
        ljmp 0xffff
__endasm;
}

void die(const char *message) {
    puts(message);
    vmstop();
}

void _assert(char *message, const char *file, unsigned int line) {
    printf("%s:%d: assertion failure: ", file, line);
    die(message);
}
