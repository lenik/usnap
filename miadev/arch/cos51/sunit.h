#ifndef COS51_SUNIT_H
#define COS51_SUNIT_H

#include <assert.h>
#include <stdio.h>
#include <8051.h>
#include "types.h"

#ifdef DEBUG
#   define TRACE(message) \
        printf("%s:%d: %s\n", __FILE__, __LINE__, message)
#else
#   define TRACE(m)
#endif

void _assert(char *message, const char *file, unsigned int line);
void die(const char *message);
void vmstop();

#ifdef SUNIT_NOCOMM
#   define SUNIT_BUFSIZE 1
#else
#   ifndef SUNIT_BUFSIZE
#   define SUNIT_BUFSIZE 32
#   endif
#endif

#ifndef SUNIT_IMPL
    __xdata byte sunitBuffer[SUNIT_BUFSIZE];
    byte sunitBufferSize = SUNIT_BUFSIZE;

    void testCase();
    void main() {
        // SCON = 0x50; // SM1, REN
        REN = 1;
        ES = 1; // IE0.4
        testCase();
        vmstop();
    }
#endif

#endif
