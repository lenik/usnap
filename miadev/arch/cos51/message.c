#include <assert.h>
#include <malloc.h>
#include <string.h>

#include "message.h"

int pollCount = 0;
const PollEntry *pollTable = NULL;
static __xdata int *pollIndexes = NULL;

int handlerCount;
const MessageHandler *handlerTable;

mbuf_t mbuf = NULL;

static int size = 0;
static int start = 0;
static int end = 0;

__xdata int statMTotal = 0;
__xdata int statMDrops = 0;

void initMQ(
        int _pollCount, const PollEntry *_pollTable,
        int _handlerCount, const MessageHandler *_handlerTable,
        int mbufSize) {
    pollCount = _pollCount;
    pollTable = _pollTable;
    pollIndexes = (__xdata int *) malloc(sizeof(int) * pollCount);
    handlerCount = _handlerCount;
    handlerTable = _handlerTable;

    size = mbufSize;
    mbuf = (mbuf_t) malloc(size);
    start = 0;
    end = 0;
}

static bool compact(int requiredSize) {
    if (end >= start) {
        int free = size - end;
        if (free < requiredSize) {
            int freesum = free + (start - 1);
            if (freesum < requiredSize)
                return false;
            else {
                memcpy(mbuf, mbuf + start, end - start);
                end -= start;
                start = 0;
            }
        }
    } else {
        int free = (start - 1) - end;
        if (free < requiredSize)
            return false;
    }
    return true;
}

void enqueue_void(m_t message) {
    statMTotal++;
    if (compact(1)) {
        mbuf[end++] = message;
        end %= size;
    } else
        statMDrops++;
}

void enqueue_b(m_t message, byte param) {
    statMTotal++;
    if (compact(2)) {
        mbuf[end++] = message;
        mbuf[end++] = param;
        end %= size;
    } else
        statMDrops++;
}

void enqueue_bb(m_t message, byte param, byte byteData) {
    statMTotal++;
    if (compact(3)) {
        mbuf[end++] = message;
        mbuf[end++] = param;
        mbuf[end++] = byteData;
        end %= size;
    } else
        statMDrops++;

}
void enqueue_w(m_t message, word data) {
    statMTotal++;
    if (compact(3)) {
        mbuf[end++] = message;
        *(word *) &mbuf[end] = data;
        end += 2;
        end %= size;
    } else
        statMDrops++;
}

void enqueue_bw(m_t message, byte param, word data) {
    statMTotal++;
    if (compact(4)) {
        mbuf[end++] = message;
        mbuf[end++] = param;
        *(word *) &mbuf[end] = data;
        end += 2;
        end %= size;
    } else
        statMDrops++;
}

/**
 * @return false if the queue is empty.
 */
bool dequeue() {
    if (end == start)
        return false;

    {
        // assert (op, param, data) in a linear space.
        m_t message = mbuf[start++];
        // FORCE-LINEAR   start %= size;

        char payloadFormat = message >> 5;
        char payloadSize = payloadFormat >> 1;
        // assert (end - start > payloadSize);

        byte param;
        word data;

        switch (payloadFormat) {
        case 0: // 00.0
        case 1: // 00.1
            param = 0;
            data = 0;
            break;
        case 2: // 01.0
        case 3: // 01.1
            param = mbuf[start++];
            data = 0;
            break;

        case 4: // 10.0
            param = mbuf[start++];
            data = mbuf[start++];
            break;

        case 5: // 10.1
            param = 0;
            data = *(word *) &mbuf[start];
            start += 2;
            break;

        case 6: // 11.0
        case 7: // 11.1
            param = mbuf[start++];
            data = *(word *) &mbuf[start];
            start += 2;
            break;

        default:
            assert(false);
            param = data = 0;
        }
        start %= size;
    }

    {
        int i;
        for (i = 0; i < handlerCount; i++) {
            MessageHandler handler = handlerTable[i];
            if ((*handler)(message, param, data))
                break;
        }
    }
    return true;
}

static bool pollingEvents = false;
void pollEvents() { // Don't have to be reentrant.

    const PollEntry *pEntry;
    __xdata int *pIndex;
    int i;

    if (pollingEvents)
        return;

    pollingEvents = 1;

    pEntry = pollTable;
    pIndex = pollIndexes;

    for (i = 0; i < pollCount; i++) {
        int interval = pEntry->pollInterval;
        int index = *pIndex;
        if (index++ == 0)
            // if (pEntry->poll()) mqDirty = true;
            pEntry->poll();
        if (index > interval)
            index = 0;
        *pIndex++ = index;
        pEntry++;
    }

    pollingEvents = 0;
}
