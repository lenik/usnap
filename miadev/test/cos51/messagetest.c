#define DEBUG
#include <cos51/sunit.h>
#include <cos51/message.h>

#define HEAP_SIZE 200
#include <_heap.h>

byte handler(m_t message, byte param, word data)
__reentrant {
    puts("Handler: ");
    printf("%d %d %d\n", message, param, data);
    return true;
}

static char val = 'a';
byte poll() {
    puts("poll!");
    if (val >= 'c')
        return 0;
    enqueue_bb(M_RECV, 7, val++);
    return 1;
}

__code PollEntry pollTables[] = { { poll, 1 }, };
__code MessageHandler handlers[] = { handler, };

void testCase() {
    TRACE("initmq");
    INITMQ(pollTables, handlers, 100);

    TR1 = 0;
    TMOD = 0x10; // 0.0.10 xxxx, Timer 1: timer, mode 2 (TH â‡’ TL)
    TH1 = 0x03; // US2TCC(100, byte);
    TL1 = 0x03;

    EA = 1;
    ET1 = 1;
    TR1 = 1;

    TRACE("dequeue-loop");
    while (val < 'h') {
        puts("dequeue");
        dequeue();
    }
}

void pollTimer()
__interrupt 3 __using 1 {
    puts("pollTimer");
    TH1 = 3;
    TL1 = 3;

    pollEvents();
}
