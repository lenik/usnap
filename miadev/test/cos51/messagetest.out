messagetest.c:28: initmq
pollTimer
poll!
Handler:
100 2 25600
pollTimer
mepollTimer
poll!
sspollTimer
agpollTimer
poll!
etpollTimer
espollTimer
poll!
t.pollTimer
cpollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
3pollTimer
poll!
341pollTimer
popollTimer
poll!
ll!pollTimer
dequpollTimer
poll!
eue
pollTimer
pollTimer
poll!
HandlerpollTimer
:
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
161pollTimer
 pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
0pollTimer
 pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
24pollTimer
839pollTimer
poll!

dpollTimer
equeue
pollTimer
poll!
pollTimer
HandpollTimer
poll!
ler:
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
16pollTimer
poll!
1 pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
pollTimer
pollTimer
poll!
0 pollTimer
pollTimer
poll!
